### PR TITLE
Webscraper Osmosis

### PR DESCRIPTION
**Pull request by ctram**
https://github.com/codeforsanjose/heartofthevalley/issues/33

This PR is for adding a scraper to gather the artwork data from [San Jose Facilities](http://sanjoseca.gov/Facilities) website.

1. Osmosis scraper package is used to iterate through all the artworks on the San Jose site, capturing details:
- artist
- title
- image source url
- postal address
- description

2. We use Nominatim to generate each artwork's geolocation.

3. We merge the scraped data with the legacy data in `art.js` (probably will remove this entirely later).

4. Save all artworks data as JSON.

5. Presently on master, the artworks data is saved within a JS file. Since it was easiest to save the scraped data as JSON, an Express server is now added so we can easily server the JSON file.

Current issues:
- Capturing details of artworks is not consistent, as the HTML structure differs from page to page. We may need humans to clean up the data within the JSON file unless another solution is found.

### CODE REVIEW

**Comment by ctram** [comment]
@ychoy @JMStudiosJoe Hello folks, some feedback please as to the direction of this PR before going too much further.

The biggest thing is the addition of an Express server to serve the scraped data, which is saved as JSON. The server serves `index.html` at `localhost:3000`. 

I started using ES6 syntax, arrow functions and const variables. I did a quick check at https://caniuse.com/#home, I think it is safe... well except for IE, hehe, but I can easily go back and revert to old syntax. Actually, yea, I'll go back and remove the arrow functions.

Don't want to step on any toes or over any conventions, so please let me know if you have comments about anything.

Ultimately I think the main challenge will be, as mentioned elsewhere, we will need to manually clean up some JSON. When I run the scraper script, it scrapes 128 pages, with 15 pages marked as errors, which means there is a missing title, artist, or creationDate (which I think we may not be displaying on the app anyway, right now).

Please feel free to come up with better rules for the scraping - otherwise, it might take less time to just manually fix up the current data.

The other challenge I see is the artist detail: I see in `art.js` we're adding links to the artist page(s) -- were those URLs found manually researched and added into `art.js`? I am thinking so because I do not see artist links on `sanjoseca.gov`.

**Comment by ctram** [comment]
@JMStudiosJoe @ychoy Hello guys, ping, am I OK to move forward with this and potentially merge? Not sure I have merge permissions, but you get the idea.

**Comment by JMStudiosJoe** [comment]
@ctram Currently this branch is able to scrape all the data we need (except for the 18)? If that is the case then yes we should merge this into master and move onto other tasks what do you think @ychoy ? Also yes this likely needs to work on IE 8 for city computers to access we need to discuss that further as well.

File: config/constants.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+const path = require('path');
+
+module.exports = {
+  MAPBOX_API_TOKEN:
+    'pk.eyJ1IjoieWNob3kiLCJhIjoiY2pmOTYwdzZ5MG52dDJ3b2JycXY4ZDU5ciJ9.m9H_Mqu1b42AObg_u_tjpA',
```
Comment (JMStudiosJoe) [inline config/constants.js]: Do we want this in the repo? might need a config file

File: public/js/main.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+ *	Heart of the Valley
+ *		main.js
+ */
+
+(async function() {
+  let res;
+  let art;
+
+  try {
+    res = await fetch('http://localhost:3000/api/artworks');
```
Comment (JMStudiosJoe) [inline public/js/main.js]: do we want to use await, might be too new. Await over promises?

File: public/js/main.js [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  if (!('remove' in Element.prototype)) {
+    Element.prototype.remove = function() {
+      if (this.parentNode) {
+        this.parentNode.removeChild(this);
+      }
+    };
+  }
+
+  mapboxgl.accessToken =
+    'pk.eyJ1IjoiY29kZWZvcmFtZXJpY2EiLCJhIjoiY2pmY3U3M3dlMXY1OTMzb2IxOGZkNW5wcyJ9.0qZCHLOM2Vy_JEE_Zx5-1g';
```
Comment (JMStudiosJoe) [inline public/js/main.js]: likely will want a config file to pull this from

File: config/constants.js (hunk 1 above)
Comment (ctram) [reply config/constants.js]: @JMStudiosJoe No, I don't think we want this in exposed in the repo. What do you mean by config file? For me, I would keep this info as a local environment variable.

File: public/js/main.js (hunk 2 above)
Comment (ctram) [reply public/js/main.js]: @JMStudiosJoe Good call, looks like IE 11 doesn't even support the native `fetch` function https://caniuse.com/#search=fetch, I'll update that.

File: config/constants.js (hunk 1 above)
Comment (JMStudiosJoe) [reply config/constants.js]: Environment variable is one way another is to have a config.json file that is ignored from hit and imported when needed to access the key along with any other keys

**Comment by ctram** [comment]
Sounds good. Those new to the project will need to get the key from us, then, no?

**Comment by JMStudiosJoe** [comment]
>Sounds good. Those new to the project will need to get the key from us, then, no? ...

@ctram yes they will need to get key from us but wouldn’t they need to do the same if they need environment variable?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ychoy** [comment]
@ctram @JMStudiosJoe - did you want to include addressing the comments about await and hiding keys inside this PR, or a separate issue? Other than those issues, I think this would be good to merge.

**Comment by ychoy** [comment]
@ctram for the 18 or so that don't work with the scraper, do we know specifically which pages they are? Maybe it's easier to just manually add that data?

**Comment by JMStudiosJoe** [comment]
Keys should be a separate issue and should not prevent the merge. I'd say good to merge too.

**Comment by ctram** [comment]
> @ctram for the 18 or so that don't work with the scraper, do we know specifically which pages they are? Maybe it's easier to just manually add that data?

@ychoy yes, currently all the artwork is scraped and written to a json file. The artworks with errors have an `error` property; the url is also saved.

Yes, at this point, it would be simpler to correct the few errors. Should be straightforward to use your editor's `find` feature and manually fix directly within the json file. I will try to do this now.

**Pull request merged**